tests/run-make/extern-fn-reachable/rmake.rs

```diff
@@ -0,0 +1,55 @@
+//! Smoke test to check that that symbols of `extern "C"` functions and `#[no_mangle]` rust
+//! functions:
+//!
+//! 1. Are externally visible in the dylib produced.
+//! 2. That the symbol visibility is orthogonal to the Rust nameres visibility of the functions
+//!    involved.
+
+//@ ignore-cross-compile
+
+use run_make_support::object::{self, Object, ObjectSection, ObjectSymbol};
+use run_make_support::{dynamic_lib_name, is_darwin, is_msvc, path, rfs, rustc};
+
+fn main() {
+    let dylib = dynamic_lib_name("dylib");
+    rustc().input("dylib.rs").output(&dylib).arg("-Cprefer-dynamic").run();
+    let mut fun1_count = 0; // #[no_mangle] pub extern "C" fn fun1() {}
+    let mut fun2_count = 0; // #[no_mangle] extern "C" fn fun2() {}
+    let mut fun3_count = 0; // mod foo { #[no_mangle] pub extern "C" fn fun3() {} }
+    let mut fun4_count = 0; // pub mod bar { #[no_mangle] pub extern "C" fn fun4() {} }
+    let mut fun5_count = 0; // #[no_mangle] pub fn fun5() {}
+    let mut fun6_count = 0; // #[no_mangle] fn fun6() {}
+
+    let p = if is_msvc() { path(dylib) } else { path(dylib) };
+    let blob = rfs::read(&p);
+    let file = object::File::parse(&*blob).unwrap();
+
+    for export in file.exports().unwrap() {
+        let sym_name = export.name();
+        let sym_name = String::from_utf8_lossy(sym_name);
+        match sym_name.as_ref() {
+            "fun1" => fun1_count += 1,
+            "fun2" => fun2_count += 1,
+            "fun3" => fun3_count += 1,
+            "fun4" => fun4_count += 1,
+            "fun5" => fun5_count += 1,
+            "fun6" => fun6_count += 1,
+            // Mach-O states that all exported symbols should have an underscore as prefix. At the
+            // same time dlsym will implicitly add it, so outside of compilers, linkers and people
+            // writing assembly, nobody needs to be aware of this.
+            "_fun1" if is_darwin() => fun1_count += 1,
+            "_fun2" if is_darwin() => fun2_count += 1,
+            "_fun3" if is_darwin() => fun3_count += 1,
+            "_fun4" if is_darwin() => fun4_count += 1,
+            "_fun5" if is_darwin() => fun5_count += 1,
+            "_fun6" if is_darwin() => fun6_count += 1,
+            _ => {}
+        }
+    }
+    assert_eq!(fun1_count, 1);
```

Comment:
as duplicate symbols feel unlikely to appear when reading a correct dylib, do you actually want the counts instead of checking the symbols are present in a set?

Reply:
I went for a stricter check because as you said, it seems unlikely that we would have duplicates. I can change this to a simple existence check if that makes more sense.

Reply:
I think it does yeah, and that should clean up the test by folding the darwin check into a per-symbol check.

Reply:
Simplified this to become an existence check, not an exact count check.